src/main/java/com/redhat/rhjmc/containerjfr/localization/LocalizationManager.java

```diff
@@ -0,0 +1,184 @@
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import com.redhat.rhjmc.containerjfr.core.log.Logger;
+
+public class LocalizationManager {
+    private Logger logger;
+
+    private Locale defaultLocale = Locale.getDefault();
+    private Map<Locale, Map<String, String>> messages = new HashMap<>();
+
+    LocalizationManager(Logger logger) {
+        this.logger = logger;
+    }
+
+    public void setDefaultLocale(Locale locale) {
+        Objects.requireNonNull(locale);
+        defaultLocale = locale;
+    }
+
+    public Locale getDefaultLocale() {
+        return defaultLocale;
+    }
+
+    public Collection<Locale> getAvailableLocales() {
+        return messages.keySet();
+    }
+
+    /**
+     * Get all messages available for default locale
+     *
+     * @return a map of message key-value pairs
+     */
+    public Map<String, String> getAllMessages() {
+        return getAllMessages(defaultLocale);
+    }
+
+    /**
+     * Get all messages available
+     *
+     * <p>If there are messages available for default locale but not for the specific locale,
+     * messages for the default locale will be used.
+     *
+     * @param locale the locale of the result messages
+     * @return a map of message key-value pairs
+     */
+    public Map<String, String> getAllMessages(Locale locale) {
+        Map<String, String> defaultMessages = messages.get(defaultLocale);
+        if (locale == defaultLocale) {
+            return new HashMap<>(defaultMessages);
+        }
+
+        if (defaultMessages == null) {
+            defaultMessages = Map.of();
+        }
+
+        Map<String, String> ret = new HashMap<>(messages.get(locale));
+        for (Map.Entry<String, String> entry : defaultMessages.entrySet()) {
+            ret.merge(
+                    entry.getKey(),
+                    entry.getValue(),
+                    (oldValue, newValue) -> {
+                        logger.warn(
+                                String.format(
+                                        "missing message for key %s in locale %s",
+                                        entry.getKey(), locale));
+                        return oldValue;
+                    });
+        }
+
+        return ret;
+    }
+
+    /**
+     * Add or overwrite a message of the given key for the default locale
+     *
+     * @param key the unique message key
+     * @param message the message content
+     */
+    public void putMessage(String key, String message) {
+        putMessage(defaultLocale, key, message);
+    }
+
+    /**
+     * Add or overwrite a message of the given key
+     *
+     * @param locale the locale of the message being added
+     * @param key the unique message key
+     * @param message the message content
+     */
+    public void putMessage(Locale locale, String key, String message) {
+        messages.putIfAbsent(locale, new HashMap<>());
```

Comment:
It would probably make sense to return a boolean indicating whether the `put` succeeded (ie value was absent).

Reply:
Sorry, I don't understand here. The put operation is always gonna success. What we have here is essentially a 2-D map, and this line it for creating a new inner map if one hasn't been created yet.